### PR TITLE
BUG: fix AttributeError on accessing object in nested MaskedArray.

### LINKED ISSUE #15895: BUG: MaskedArray with nested dtype and object elements cause AttributeError on access
When I create a `MaskedArray` using a nested dtype which contains an `object` at the innermost level, trying to access it causes an _AttributeError: 'str' object has no attribute 'ndim'_.

If I use `int` instead of `object`, the element access works fine.

The issue seems to be caused by the assumption that `_fill_value` is of a numpy type and has an `ndim` attribute.
This assumption is true for `numpy.int32` which would be used if `int` was specified inside the dtype.
But for `object`, the `_fill_value` will be a just python `str` _'?'_, which has no attribute `ndim`.

I propose to change the code line in question to
```python
if hasattr(dout._fill_value, 'ndim') and dout._fill_value.ndim > 0:
```

### Reproducing code example:

```python
import numpy as np
my_dtype = np.dtype([('b', [('c', object)], (1,))])
a = np.ma.masked_all((1,), my_dtype)
print(a['b']['c'])
```

### Error message:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Python37\lib\site-packages\numpy\ma\core.py", line 3276, in __getitem__
    if dout._fill_value.ndim > 0:
AttributeError: 'str' object has no attribute 'ndim'
```

### Numpy/Python version information:

1.18.1 3.7.4 (tags/v3.7.4:e09359112e, Jul  8 2019, 20:34:20) [MSC v.1916 64 bit (AMD64)]

### PR DESCRIPTION
**Pull request by rena-xu**
Closes #15895

### CODE REVIEW

**Comment by rena-xu** [comment]
Sorry I'm really new to Travis-CI. I tried to check the details of the run but it seems like everything passed. I'm wondering why the check is still considered pending? Is there anything I could fix here? Thanks!

File: numpy/ma/core.py [hunk 1]
```diff
@@ -3316,6 +3317,11 @@ def _scalar_heuristic(arr, elem):
                                 f"all to {dout._fill_value[0]!s}.",
                                 stacklevel=2)
                         dout._fill_value = dout._fill_value.flat[0]
+                    elif not hasattr(dout._fill_value, 'ndim'):
+                        # only initialize to empty array if it dout does not 
+                        # have ._fill_value before. If it has, do not change 
+                        # the original behavior
+                        dout._fill_value = np.array([], dtype)
```
Comment (eric-wieser) [inline numpy/ma/core.py:3324]: This doesn't strike me as correct, `_fill_value` should always be a 0d array, this is a 1d array.

Comment (rena-xu) [reply numpy/ma/core.py:3324]: Oops sorry I'll look into that!

File: numpy/ma/core.py [hunk 2]
```diff
@@ -3316,6 +3317,11 @@ def _scalar_heuristic(arr, elem):
                                 f"all to {dout._fill_value[0]!s}.",
                                 stacklevel=2)
                         dout._fill_value = dout._fill_value.flat[0]
```
Comment (eric-wieser) [inline numpy/ma/core.py:3319]: This line looks like it has a bug too, sadly. I'll see if I can produce a test-case at some point.

File: numpy/ma/core.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/ma/core.py:3324]: I suspect you want
```suggestion
                        dout._fill_value = np.empty((), dtype)
                        dout._fill_value[()] = self._fill_value[indx]
```

Comment (rena-xu) [reply numpy/ma/core.py:3324]: Yes thank you! I'll try to add some tests for this in the next commit!

File: numpy/ma/tests/test_extras.py [hunk 3]
```diff
@@ -74,6 +74,7 @@ def test_masked_all_with_dtype_str(self):
         assert_equal(type(a['b']['c']), np.ma.core.MaskedArray)
         assert_equal(len(a['b']['c']), 1)
         assert_equal(a['b']['c'].shape, (1, 1))
+        assert_equal(a['b']['c']._fill_value.shape, np.empty((), object))
```
Comment (eric-wieser) [inline numpy/ma/tests/test_extras.py:77]:
```suggestion
        assert_equal(a['b']['c']._fill_value.shape, ())
``

File: numpy/ma/tests/test_extras.py [hunk 4]
```diff
@@ -64,6 +64,18 @@ def test_masked_all(self):
         control = array([[(1, (1, 1))]], mask=[[(1, (1, 1))]], dtype=dt)
         assert_equal(test, control)
 
+    def test_masked_all_with_dtype_str(self):
+        # Test masked_all works with nested array with dtype of an 'object'
+        # refers to issue #15895
+        my_dtype = np.dtype([('b', [('c', object)], (1,))])
```
Comment (eric-wieser) [inline numpy/ma/tests/test_extras.py:70]: The other issue case I identified above is when the dtype is `my_dtype = np.dtype([('b', (object, (1,)))])`

Comment (eric-wieser) [inline numpy/ma/tests/test_extras.py:70]: Possibly clearer as:
```suggestion
        my_dtype = np.dtype([('b', ([('c', object)], (1,)))])
```
or

```suggestion
        struct_with_object = np.dtype([('c', object)])
        my_dtype = np.dtype([('b', (struct_with_object, (1,)))])
```

Comment (rena-xu) [reply numpy/ma/tests/test_extras.py]: Just to make sure, if we try to access arr['b'], the expected _fill_value for it is still () right?

File: numpy/ma/core.py [hunk 5]
```diff
@@ -3316,6 +3317,12 @@ def _scalar_heuristic(arr, elem):
...
+                    elif not hasattr(dout._fill_value, 'ndim'):
+                        # only initialize to empty array if it dout does not 
+                        # have ._fill_value before. If it has, do not change 
+                        # the original behavior
+                        dout._fill_value = np.empty((), dtype)
+                        dout._fill_value[()] = self._fill_value[indx]
```
Comment (eric-wieser) [inline numpy/ma/core.py:3325]: I suspect that actually the fix is way simpler, and that you can get away with the only change being
```suggestion
                        # need to use squeeze to ensure the result is an array not a scalar
                        dout._fill_value = dout._fill_value.flat[0:1].squeeze(axis=0)
```

Would you mind testing that?

Comment (rena-xu) [reply numpy/ma/core.py:3325]: Oh yes I've tested it and it works! Thank you!

Comment (rena-xu) [reply numpy/ma/core.py:3325]: I'll pushed the tests along with the change

Comment (eric-wieser) [reply numpy/ma/core.py:3325]: Thanks for taking the time to write the tests so that it's easy to try these things!

File: numpy/ma/core.py [hunk 6]
```diff
@@ -3316,13 +3316,10 @@ def _scalar_heuristic(arr, elem):
...
-                    elif not hasattr(dout._fill_value, 'ndim'):
-                        # only initialize to empty array if it dout does not 
-                        # have ._fill_value before. If it has, do not change 
-                        # the original behavior
-                        dout._fill_value = np.empty((), dtype)
-                        dout._fill_value[()] = self._fill_value[indx]
+                    # need to use squeeze to ensure the result is an array 
+                    # not a scalar. If the _fill_value is like a str type,
+                    # then just make sure we turn the _fill_value to a 0d array
+                    dout._fill_value = dout._fill_value.flat[0:1].squeeze(axis=0)
```
Comment (eric-wieser) [inline numpy/ma/core.py:3322]: Needs indenting by one

Comment (rena-xu) [reply numpy/ma/core.py:3322]: But if we indent by one, then it does not solve for the case when dtype.ndim is 0 though

Comment (eric-wieser) [reply numpy/ma/core.py:3322]: I don't believe the problem exists when it's 0

Comment (eric-wieser) [reply numpy/ma/core.py:3322]: The problem was caused by the use of `.flat`

Comment (rena-xu) [reply numpy/ma/core.py:3322]: Ah I see! Thank you!

**Comment by rena-xu** [comment]
Hi this is really my first time contributing to an open-source project! I'm wondering if it's possible for these changes to be accepted? If the code and tests are not good enough I'll definitely try to modify something to it!

File: numpy/ma/core.py [hunk 7]
```diff
@@ -3305,7 +3305,8 @@ def _scalar_heuristic(arr, elem):
                     # which should have shape () to avoid breaking several
                     # methods. There is no great way out, so set to
                     # first element.  See issue #6723.
-                    if dout._fill_value.ndim > 0:
+                    dtype = self._fill_value.dtype[indx]
+                    if dtype.ndim > 0:
```
Comment (eric-wieser) [inline numpy/ma/core.py:3309]: Is this change necessary?

I'm tempted to add `assert isinstance(self._fill_value, np.ndarray) before this line.

Comment (rena-xu) [reply numpy/ma/core.py:3309]: Sorry I'm kinda confused here, so if the _fill_value is a string instead of ndarray, are we suppose to throw an error or change it to a 0darray?

Comment (eric-wieser) [reply numpy/ma/core.py:3309]: `_fill_value` should never be anything other than a 0d array. If it's somehow a string here, then something else has a bug, and we should track down that bug.

Comment (eric-wieser) [inline numpy/ma/core.py:3309]: To be clear, I'd suggest we change these lines to
```suggestion
                    # something like gh-15895 has happened if this fails
                    assert isinstance(dout._fill_value, np.ndarray)
                    if dout._fill_value.ndim > 0:
```

Comment (rena-xu) [reply numpy/ma/core.py:3309]: Ah I see, guess I was not understanding the expected behavior correctly.

**Comment by rgommers** [comment]
> Hi this is really my first time contributing to an open-source project! I'm wondering if it's possible for these changes to be accepted? If the code and tests are not good enough I'll definitely try to modify something to it!

Hi @rena-xu, welcome! You're doing everything right, no worries. Review typically takes a while, looks like this one is getting close to merging.

File: numpy/ma/core.py [hunk 8]
```diff
@@ -3305,6 +3305,8 @@ def _scalar_heuristic(arr, elem):
                     # which should have shape () to avoid breaking several
                     # methods. There is no great way out, so set to
                     # first element.  See issue #6723.
+                    # something like gh-15895 has happened if this fails
+                    assert isinstance(dout._fill_value, np.ndarray)
```
Comment (mattip) [inline numpy/ma/core.py:3309]: This seems wrong. If there is a way for `dout._fill_value` to ever not be an `ndarray`, then this should raise a `ValueError`, not an `AssertionError`. And it would be even better if there was a test to show how this could ever happen.

Comment (eric-wieser) [reply numpy/ma/core.py:3309]: > it would be even better if there was a test to show how this could ever happen.

It can't ever happen, as far as a I know - but this code is bug-prone, and this makes it easier for us to work out the issue if it does fire.

We could use `RuntimeError("internal numpy error")` or similar if you prefer.

Comment (mattip) [reply numpy/ma/core.py:3309]: +1. As far as I can tell we have no other asserts in the code, we do have RuntimeErrors for this kind of thing.

**Comment by vrakesh** [comment]
@rena-xu Requesting an update on this. Changes to assertion as suggested above should move this forward

**Comment by rena-xu** [comment]
@vrakesh Sorry for the delay! I'll change this ASAP.

**Comment by rena-xu** [comment]
Sorry I'm kinda new to this. I'm wondering if it's possible to re-run the azure-pipeline test? It failed due to an ImportError, which I feel like should not be cause by this code change. I was able to pass  this part by running runtests.py locally, so I'm not sure if I can reproduce this error

**Comment by rgommers** [comment]
> I'm wondering if it's possible to re-run the azure-pipeline test? It failed due to an ImportError, which I feel like should not be cause by this code change

Yes, you're right - that failure is happening on other PRs as well, re-running isn't going to fix it right now. We're fine ignoring unrelated failures, so you're all good here. Time for a re-review of your most recent changes.

File: numpy/ma/core.py [hunk 9]
```diff
@@ -3305,6 +3305,10 @@ def _scalar_heuristic(arr, elem):
                     # which should have shape () to avoid breaking several
                     # methods. There is no great way out, so set to
                     # first element.  See issue #6723.
+                    # something like gh-15895 has happened if this check fails.
+                    # _fill_value should always be an ndarray.
+                    if not isinstance(dout._fill_value, np.ndarray):
+                        raise RuntimeError('Internal NumPy error.')
```
Comment (eric-wieser) [inline numpy/ma/core.py:3311]: nit: can you move these lines up to line 3299? Where you've put them has separated the comment above from the code it describes.

File: numpy/ma/core.py [hunk 10]
```diff
@@ -3315,7 +3319,10 @@ def _scalar_heuristic(arr, elem):
                                 "heterogeneous fill_value and setting "
                                 f"all to {dout._fill_value[0]!s}.",
                                 stacklevel=2)
-                        dout._fill_value = dout._fill_value.flat[0]
+                        # need to use squeeze to ensure the result is an array 
+                        # not a scalar. If the _fill_value is like a str type,
+                        # then just make sure we turn the _fill_value to a 0d array
+                        dout._fill_value = dout._fill_value.flat[0:1].squeeze(axis=0)
```
Comment (eric-wieser) [inline numpy/ma/core.py:3325]: This doesn't quite capture the problem here. Perhaps
```suggestion
                        # Need to use `.flat[0:1].squeeze(...)` instead of just
                        # `.flat[0]` to ensure the result is a 0d array and not
                        # a scalar.
                        dout._fill_value = dout._fill_value.flat[0:1].squeeze(axis=0)
```

File: numpy/ma/tests/test_extras.py [hunk 11]
```diff
@@ -64,6 +64,28 @@ def test_masked_all(self):
...
+        masked_arr = np.ma.masked_all((1,), my_dtype)
+
+        assert_equal(type(masked_arr['b']), np.ma.core.MaskedArray)
+        assert_equal(type(masked_arr['b']['c']), np.ma.core.MaskedArray)
+        assert_equal(len(masked_arr['b']['c']), 1)
+        assert_equal(masked_arr['b']['c'].shape, (1, 1))
+        assert_equal(masked_arr['b']['c']._fill_value.shape, ())
+    
+    def test_masked_all_with_object(self):
+        # same test above except that the array si not nested
```
Comment (eric-wieser) [inline numpy/ma/tests/test_extras.py:80]:
```suggestion
        # same as above except that the array is not nested
```

**Review by eric-wieser — APPROVED** [review]
Looks good now, thanks for persevering!
